hydra-node/test/Hydra/Model.hs

```diff
@@ -0,0 +1,579 @@
+            pendingCommits' = party `Set.delete` pendingCommits
+            updatedState =
+              if null pendingCommits'
+                then
+                  Open
+                    { headParameters
+                    , offChainState =
+                        OffChainState
+                          { confirmedUTxO = mconcat (Map.elems commits')
+                          , seenTransactions = []
+                          }
+                    }
+                else
+                  Initial
+                    { headParameters
+                    , commits = commits'
+                    , pendingCommits = pendingCommits'
+                    }
+          _ -> error "unexpected state"
+      --
+      Command{command = Input.Abort} ->
+        WorldState{hydraParties, hydraState = updateWithAbort hydraState}
+       where
+        updateWithAbort = \case
+          Initial{commits} -> Final committedUTxO
+           where
+            committedUTxO = mconcat $ Map.elems commits
+          _ -> Final mempty
+      --
+      Command{command = Input.NewTx{Input.transaction = tx}} ->
+        WorldState{hydraParties, hydraState = updateWithNewTx hydraState}
+       where
+        updateWithNewTx = \case
+          Open{headParameters, offChainState = OffChainState{confirmedUTxO, seenTransactions}} ->
+            Open
+              { headParameters
+              , offChainState =
+                  OffChainState
+                    { confirmedUTxO = confirmedUTxO `applyTx` tx
+                    , seenTransactions = tx : seenTransactions
+                    }
+              }
+          _ -> error "unexpected state"
+      _ -> error "not implemented"
+
+  perform :: WorldState m -> Action (WorldState m) a -> LookUp -> StateT (Nodes m) m a
+  perform _ Seed{seedKeys} _ = seedWorld seedKeys
+  perform st Command{party, command} _ = do
+    case command of
+      Input.Commit{Input.utxo = utxo} ->
+        performCommit party utxo
+      Input.NewTx{Input.transaction = tx} ->
+        performNewTx st party tx
+      Input.Init{Input.contestationPeriod = p} ->
+        party `performs` Input.Init{Input.contestationPeriod = p}
+      Input.Abort -> do
+        party `performs` Input.Abort
+      Input.GetUTxO -> do
+        party `performs` Input.GetUTxO
+      Input.Close -> do
+        party `performs` Input.Close
+      Input.Contest -> do
+        party `performs` Input.Contest
+      Input.Fanout -> do
+        party `performs` Input.Fanout
+
+  monitoring (s, s') _action _lookup _return =
+    case (hydraState s, hydraState s') of
+      (st, st') -> tabulate "Transitions" [unsafeConstructorName st <> " -> " <> unsafeConstructorName st']
+
+deriving instance Show (Action (WorldState m) a)
+deriving instance Eq (Action (WorldState m) a)
+
+--
+
+-- * Perform Helpers
+
+--
+
+performs :: Monad m => Party -> ClientInput Tx -> StateT (Nodes m) m ()
+performs party command = do
```

Comment:
Could: I was confusing this with the frameworks' `perform` function. Maybe `sendsInput` is a better name?